pallets/parachain-staking/src/tests.rs

```diff
@@ -9549,3 +9549,67 @@ fn test_delegate_skips_auto_compound_storage_but_emits_event_for_zero_auto_compo
 			}));
 		});
 }
+
+#[test]
+fn test_on_initialize_weights() {
+	use crate::mock::System;
+	use crate::weights::{SubstrateWeight as PalletWeights, WeightInfo};
+	use crate::*;
+	use frame_support::{pallet_prelude::*, weights::constants::RocksDbWeight};
+
+	// generate balance, candidate, and delegation vecs to "fill" out delegations
+	let mut balances = Vec::new();
+	let mut candidates = Vec::new();
+	let mut delegations = Vec::new();
+
+	for collator in 1..30 {
+		balances.push((collator, 100));
+		candidates.push((collator, 10));
+		let starting_delegator = collator * 1000;
+		for delegator in starting_delegator..starting_delegator + 300 {
+			balances.push((delegator, 100));
+			delegations.push((delegator, collator, 10));
+		}
+	}
+
+	ExtBuilder::default()
+		.with_balances(balances)
+		.with_candidates(candidates)
+		.with_delegations(delegations)
+		.build()
+		.execute_with(|| {
+			let weight = ParachainStaking::on_initialize(1);
+
+			// TODO: build this with proper db reads/writes
+			assert_eq!(Weight::from_ref_time(286_002_000), weight);
+
+			// roll to the end of the round, then run on_init again, we should see round change...
+			roll_to_round_end(3);
+			set_author(2, 1, 100); // must set some points for prepare_staking_payouts
+			let block = System::block_number() + 1;
+			let weight = ParachainStaking::on_initialize(block);
+			assert_eq!(Weight::from_ref_time(3_017_871_000), weight);
```

Comment:
How is this magic number derived?

Reply:
I think most of these magic values come from the benchmarked `weights.rs`. I should add a note about that...

Reply:
Added some notes about this particular value in 690ab08